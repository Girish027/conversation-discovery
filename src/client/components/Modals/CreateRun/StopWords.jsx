import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';

import {
  TextField, Plus, Xmark, Button
} from '@tfs/ui-components';
import IconButton from '../../IconButton/IconButton';
import { stopwordsStyle } from '../../../styles/index';
import { shouldHandle } from '../../../utils/KeyboardUtils';
import GenericTooltip from '../../Tooltip/Tooltip';
import Language from '../../../Language';

export default class StopWords extends PureComponent {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.onClickBack = this.onClickBack.bind(this);

    this.state = {
      inputValue: '',
      stopWords: []
    };
    this.backButton = ' << Back ';
  }

    onEnterPressed = (event) => {
      if (shouldHandle(event)) {
        this.addStopWord(event.target.value);
      }
    }

    onClickBack() {
      this.props.onClickBack();
    }

    addStopWord = (stopWord) => {
      const { stopWords } = this.state;
      let newList = stopWords.slice();

      stopWord = stopWord.trim();

      if (!(stopWords.indexOf(stopWord) > -1) && stopWord.length > 0) {
        newList = newList.concat([stopWord]);
        this.updateStopWords(newList);
      }

      this.updateValue('');
    }

    updateValue = (value) => {
      if (value === ' ') {
        return;
      }
      this.setState({
        inputValue: value
      });
    }


    removeStopWord = (removeStopWord) => {
      const { stopWords } = this.state;
      const newStopWords = stopWords.filter((tag) => tag !== removeStopWord);
      this.updateStopWords(newStopWords);
    }

    updateStopWords = (stopWords) => {
      this.setState({
        stopWords
      }, () => this.props.onStopWordChange(stopWords));
    }

    handleChange(e) {
      this.updateValue(e.target.value);
    }

    render() {
      const { inputValue, stopWords } = this.state;

      return (
        <React.Fragment>
          <div style={stopwordsStyle.label}>
            <span> Watch Words </span>
            <Button
              name='back-btn'
              type='flat'
              onClick={this.onClickBack}
              styleOverride={{ float: 'right' }}
            >
              {this.backButton}
            </Button>
          </div>
          <div style={stopwordsStyle.filedIcon}>
            <IconButton
              onClick={() => this.addStopWord(inputValue)}
              icon={Plus}
              strokeColor='#ffffff'
              data-qa={'add-stopword'}
              padding={'3px'}
              title='Add'
            />
          </div>
          <GenericTooltip content={Language.RUN.TOOLTIP.watch}>
            <TextField
              name='stop-words'
              onChange={this.handleChange}
              value={inputValue}
              placeholder='Add a new watch word'
              styleOverride={stopwordsStyle.field}
              onKeyPress={(e) => this.onEnterPressed(e)}
            />
          </GenericTooltip>
          <div style={stopwordsStyle.stopWordsWrap}>
            {stopWords && stopWords.map((tag) => (
              <div
                style={stopwordsStyle.tagStyle}
                key={tag}
                label={tag}
              >
                <div style={stopwordsStyle.pillStyle}>
                  {tag}
                  <IconButton
                    onClick={() => this.removeStopWord(tag)}
                    icon={Xmark}
                    data-qa={'remove-stopword'}
                    padding={'0px 0px 0px 14px'}
                    title='Remove'
                  />
                </div>
              </div>
            ))}
          </div>
        </React.Fragment>
      );
    }
}

StopWords.propTypes = {
  onStopWordChange: PropTypes.func,
  onClickBack: PropTypes.func,
};

StopWords.defaultProps = {
  onStopWordChange: () => {},
  onClickBack: () => {},
};
