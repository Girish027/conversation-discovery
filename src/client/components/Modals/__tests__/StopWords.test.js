import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import StopWords from '../CreateRun/StopWords';

describe('<StopWords />', function () {
  let wrapper;


  const props = {
    errorMsg: 'Project ABC already exist',
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  const getShallowWrapper = (propsObj) => shallow(<StopWords
    {...propsObj}
  />);

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = getShallowWrapper(props);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly with default props', () => {
      wrapper = getShallowWrapper(props);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality:', () => {
    beforeEach(() => {
      props.onClickBack = jest.fn();
      wrapper = getShallowWrapper(props);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    describe('addStopWord', () => {
      test('should add the stop words', () => {
        wrapper.instance().addStopWord('Hello');
        expect(wrapper.state().stopWords).toEqual(['Hello']);
      });

      test('should not add the stop words if it is already exist', () => {
        wrapper.instance().addStopWord('Hello');
        expect(wrapper.state().stopWords).toEqual(['Hello']);
        wrapper.instance().addStopWord('World');
        expect(wrapper.state().stopWords).toEqual(['Hello', 'World']);
        wrapper.instance().addStopWord('Hello');
        expect(wrapper.state().stopWords).toEqual(['Hello', 'World']);
      });
    });

    describe('updateValue', () => {
      test('should update the input value', () => {
        wrapper.instance().updateValue('Hello');
        expect(wrapper.state().inputValue).toEqual('Hello');
      });

      test('should not update the input value when space is entered', () => {
        wrapper.instance().updateValue(' ');
        expect(wrapper.state().inputValue).toEqual('');
      });
    });

    describe('removeStopWord', () => {
      test('should remove the stop words from the list', () => {
        wrapper.instance().addStopWord('Hello');
        wrapper.instance().addStopWord('World');
        expect(wrapper.state().stopWords).toEqual(['Hello', 'World']);

        wrapper.instance().removeStopWord('Hello');
        expect(wrapper.state().stopWords).toEqual(['World']);
      });
    });

    describe('handleChange', function () {
      test('should pass the props and change the state with new value in the form', () => {
        wrapper.instance().handleChange({
          target: {
            value: 'New Stop Word'
          }
        });
        expect(wrapper.state().inputValue).toEqual('New Stop Word');
      });
    });

    describe('handleBack', function () {
      test('should go back to form page', () => {
        wrapper.instance().onClickBack();
        expect(props.onClickBack).toHaveBeenCalled();
      });
    });

    describe('onEnterPressed', function () {
      test('should add the stopword when enter is pressed', () => {
        const event = {
          target: {
            value: 'New Stop Word'
          },
          keyCode: 13,
        };
        wrapper.instance().onEnterPressed(event);
        expect(wrapper.state().stopWords).toEqual(['New Stop Word']);
      });
    });
  });
});
