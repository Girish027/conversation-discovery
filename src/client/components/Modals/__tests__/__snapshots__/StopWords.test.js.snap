// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StopWords /> Snapshots renders correctly with default props 1`] = `
<Fragment>
  <div
    style={
      Object {
        "color": "#6E6E6E",
        "fontSize": "14px",
        "letterSpacing": 0,
        "marginTop": "20px",
        "paddingBottom": "10px",
        "textAlign": "left",
      }
    }
  >
    <span>
       Watch Words 
    </span>
    <withTheme(Button)
      name="back-btn"
      onClick={[Function]}
      styleOverride={
        Object {
          "float": "right",
        }
      }
      type="flat"
    >
       &lt;&lt; Back 
    </withTheme(Button)>
  </div>
  <div
    style={
      Object {
        "padding": "10px",
        "position": "absolute",
        "right": 0,
        "zIndex": 1,
      }
    }
  >
    <IconButton
      data-qa="add-stopword"
      defaultColor="#004c97"
      disabled={false}
      disabledColor="#9B9B9B"
      focusedColor="#003467"
      height={10}
      icon={[Function]}
      onClick={[Function]}
      padding="3px"
      strokeColor="#ffffff"
      title="Add"
      width={10}
    />
  </div>
  <GenericTooltip
    content="In the natural language process, insignificant words (such as \\"of\\", \\"the\\" etc) are referred to as Stop Words. These stop words are generally ignored while creating an analysis. Keywords that need to be considered for clustering and have to be exempted from being treated as ’Stop Words’ are provided here."
    direction="bottom"
    type="info"
  >
    <withTheme(TextField)
      name="stop-words"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Add a new watch word"
      styleOverride={
        Object {
          "background": "#FFFFFF",
          "border": "0px solid #DCE1E8",
          "borderRadius": "2px",
          "input": Object {
            "fontSize": "14px",
            "letterSpacing": 0,
            "textAlign": "left",
          },
          "width": "240%",
        }
      }
      value=""
    />
  </GenericTooltip>
  <div
    style={
      Object {
        "display": "flex",
        "flexWrap": "wrap",
        "marginTop": "15px",
      }
    }
  />
</Fragment>
`;
